logging/env_logger.h

```diff
@@ -0,0 +1,177 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+//
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+//
+// Logger implementation that uses custom Env object for logging.
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include "port/sys_time.h"
+#include <time.h>
+
+#include "monitoring/iostats_context_imp.h"
+#include "rocksdb/env.h"
+#include "rocksdb/slice.h"
+#include "test_util/sync_point.h"
+#include "util/file_reader_writer.h"
+#include "util/mutexlock.h"
+
+namespace rocksdb {
+
+class EnvLogger : public Logger {
+ public:
+  EnvLogger(std::unique_ptr<WritableFileWriter>&& file, uint64_t(gettid)(),
+            Env* env, InfoLogLevel log_level = InfoLogLevel::ERROR_LEVEL)
+      : Logger(log_level),
+        file_(std::move(file)),
+        gettid_(gettid),
+        last_flush_micros_(0),
+        env_(env),
+        flush_pending_(false) {}
+
+  EnvLogger(const std::string& fname, const EnvOptions& options,
+            uint64_t(gettid)(), Env* env,
+            InfoLogLevel log_level = InfoLogLevel::ERROR_LEVEL)
+      : Logger(log_level),
+        gettid_(gettid),
+        last_flush_micros_(0),
+        env_(env),
+        flush_pending_(false) {
+    std::unique_ptr<WritableFile> writable_file;
+    env_->NewWritableFile(fname, &writable_file, options);
+    file_ = std::unique_ptr<WritableFileWriter>(
```

Comment:
Why a pointer, not a member of `WritableFileWriter` in EnvLogger?

Reply:
Siying,

This comment is not clear to me. Can you please clarify? 
Do you mean that 'file_' should be declared as WritableFileWriter instead of std::unique_ptr<WritableFileWriter>?

Reason i used unique_ptr is because the first constructor takes an unique ptr to WritableFileWriter to ensure that this class has exclusive access to the WritableFileWriter. I can change it to take a rvalue reference instead.

Reply:
Yes, that is exactly what i mean: cleared as `WritableFileWriter` instead of std::unique_ptr.

Can you explain why a constructor needs to take WritableFileWriter as a parameter, instead of WriteableFile?

Reply:
Making the change to remove unique_ptr and to take WriteableFile as param in the first constructor.

Reply:
I will remove the second constructor as the factory method Env::NewLogger() can be used instead to achieve the same.